src/base/bittorrent/session.cpp

```diff
@@ -3304,14 +3320,37 @@ bool Session::isListening() const
     return m_nativeSession->is_listening();
 }
 
-MaxRatioAction Session::maxRatioAction() const
+int Session::maxRatioActionIndex() const
```

Comment:
Sorry, but these changes totally incorrect. Core component should know nothing about any top level components and their logic. BitTorrent::Session should operate with MaxRatioAction. All mapping should be done where it's needed.

Reply:
Any appropriate place to make functions for them or i just duplicate the code in optionsdialog and appcontroller?

Reply:
You don't need to change appcontroller, isn't it? It still should return serialized MaxRatioAction.

Reply:
>It still should return serialized MaxRatioAction.

Sorry don't know what that means. I mean the following (and the equivalent in optionsdialog)
<details><summary>diff</summary>

```
--- a/src/base/bittorrent/session.cpp
+++ b/src/base/bittorrent/session.cpp
@@ -3320,37 +3323,23 @@ bool Session::isListening() const
     return m_nativeSession->is_listening();
 }
 
-int Session::maxRatioActionIndex() const
+MaxRatioAction Session::maxRatioAction() const
 {
-    switch (m_maxRatioAction) {
-    case Pause: return 0;
-    case Remove: return 1;
-    case DeleteFiles: return 2;
-    case DeleteFolder: return 3;
-    case EnableSuperSeeding: return 4;
-    default: return 0;
-    default: return 0;
-    }
+    return static_cast<MaxRatioAction>(m_maxRatioAction.value());
 }
 
-void Session::setMaxRatioAction(const int comboIndex)
+void Session::setMaxRatioAction(const MaxRatioAction act)
 {
-    switch (comboIndex) {
-    case 0:
-    default:
-        m_maxRatioAction = Pause;
-        break;
-    case 1:
-        m_maxRatioAction = Remove;
-        break;
-    case 2:
-        m_maxRatioAction = DeleteFiles;
-        break;
-    case 3:
-        m_maxRatioAction = DeleteFolder;
-        break;
-    case 4:
-        m_maxRatioAction = EnableSuperSeeding;
-    }
+    m_maxRatioAction = static_cast<int>(act);
 }
 
 // If this functions returns true, we cannot add torrent to session,
--- a/src/webui/api/appcontroller.cpp
+++ b/src/webui/api/appcontroller.cpp
@@ -208,7 +208,27 @@ void AppController::preferencesAction()
     data["max_ratio"] = session->globalMaxRatio();
     data["max_seeding_time_enabled"] = (session->globalMaxSeedingMinutes() >= 0.);
     data["max_seeding_time"] = session->globalMaxSeedingMinutes();
-    data["max_ratio_act"] = session->maxRatioActionIndex();
+
+    const MaxRatioAction maxRatioAction = session->maxRatioAction();
+    int index = 0;
+    switch (maxRatioAction) {
+    case Pause:
+        index = 0;
+        break;
+    case Remove:
+        index = 1;
+        break;
+    case DeleteFiles:
+        index = 2;
+        break;
+    case DeleteRootFolder:
+        index = 3;
+        break;
+    case EnableSuperSeeding:
+        index = 4;
+        break;
+    }
+
+    data["max_ratio_act"] = index;
     // Add trackers
     data["add_trackers_enabled"] = session->isAddTrackersEnabled();
     data["add_trackers"] = session->additionalTrackers();
@@ -547,8 +567,29 @@ void AppController::setPreferencesAction()
         else
             session->setGlobalMaxSeedingMinutes(-1);
     }
-    if (hasKey("max_ratio_act"))
-        session->setMaxRatioAction(it.value().toInt());
+    if (hasKey("max_ratio_act")) {
+        const int comboCurrentIndex = it.value().toInt();
+        MaxRatioAction maxRatioAction;
+        switch (comboCurrentIndex) {
+        case 0:
+        default:
+            maxRatioAction = Pause;
+            break;
+        case 1:
+            maxRatioAction = Remove;
+            break;
+        case 2:
+            maxRatioAction = DeleteFiles;
+            break;
+        case 3:
+            maxRatioAction = DeleteRootFolder;
+            break;
+        case 4:
+            maxRatioAction = EnableSuperSeeding;
+            break;
+        }
+
+        session->setMaxRatioAction(maxRatioAction);
+    }
     // Add trackers
     session->setAddTrackersEnabled(m["add_trackers_enabled"].toBool());
     session->setAdditionalTrackers(m["add_trackers"].toString());

```
</details>

Reply:
>Sorry don't know what that means.

Just don't touch AppController at all.

Reply:
Also you shouldn't change maxRatioAction/setMaxRatioAction.

Reply:
What? I reverted maxRatioAction/setMaxRatioAction but if i don't do anything in appcontroller it still saves/loads based on index. Where am i supposed to do that mapping if not there?

Reply:
Well, let's start from the beginning:
1. You need to add some new MaxRatioAction values. Just add it there. You don't need to change anything else (except, of course, the logic that uses these new behaviors). maxRatioAction/setMaxRatioAction is still correct (if you didn't change internal values of existing MaxRatioAction values). Also AppController is still valid since it send expected values for previously existing MaxRatioActions.
2. You need to add appropriate options in GUI/WebUI. Previously they use simplified way of using internal MaxRatioAction representation. Now you don't want to use it since it causes unwanted order of options in appropriate controls. So you need to define some mapping between MaxRatioAction serialized representation values and UI representation values (in GUI and WebUI).

Reply:
In fact, we often use some more or less simplified methods. If you like, I can describe this part of the logic to you from a General point of view.

Reply:
>In fact, we often use some more or less simplified methods. 

Could you point me to a few of them?

>If you like, I can describe this part of the logic to you from a General point of view.

Please do.

Reply:
>So you need to define some mapping between MaxRatioAction serialized representation values and UI representation values (in GUI and WebUI).

If the following is correct for the GUI, where do i do it for the WebUI?
<details><summary>diff</summary>

```
--- a/src/gui/optionsdialog.cpp
+++ b/src/gui/optionsdialog.cpp
@@ -747,7 +747,15 @@ void OptionsDialog::saveOptions()
     session->setAdditionalTrackers(m_ui->textTrackers->toPlainText());
     session->setGlobalMaxRatio(getMaxRatio());
     session->setGlobalMaxSeedingMinutes(getMaxSeedingMinutes());
-    session->setMaxRatioAction(static_cast<MaxRatioAction>(m_ui->comboRatioLimitAct->currentIndex()));
+
+    const QHash <int, MaxRatioAction> actIndex = {
+        {0, Pause},
+        {1, Remove},
+        {2, DeleteFiles},
+        {3, DeleteRootFolder},
+        {4, EnableSuperSeeding}
+    };
+    session->setMaxRatioAction(actIndex.value(m_ui->comboRatioLimitAct->currentIndex()));
     // End Bittorrent preferences
 
     // Misc preferences
@@ -1130,7 +1138,15 @@ void OptionsDialog::loadOptions()
         m_ui->spinMaxSeedingMinutes->setEnabled(false);
     }
     m_ui->comboRatioLimitAct->setEnabled((session->globalMaxSeedingMinutes() >= 0) || (session->globalMaxRatio() >= 0.));
-    m_ui->comboRatioLimitAct->setCurrentIndex(session->maxRatioActionIndex());
+
+    const QHash <int, MaxRatioAction> actIndex = {
+        {0, Pause},
+        {1, Remove},
+        {2, DeleteFiles},
+        {3, DeleteRootFolder},
+        {4, EnableSuperSeeding}
+    };
+    m_ui->comboRatioLimitAct->setCurrentIndex(actIndex.key(session->maxRatioAction()));
     // End Bittorrent preferences
 
     // Web UI preferences
```
</details>

Reply:
>If the following is correct for the GUI

The first mapping is correct. The second one should be `QHash<MaxRatioAction, int>`. Lookup for key is slow.

Reply:
>where do i do it for the WebUI?

Just assign appropriate values to "option" elements. Isn't it enough?

Reply:
>Just assign appropriate values to "option" elements.

OMG this was the first thing i tried but i thought it didn't work... guess i forgot to restart qBt after compile. Anyway, what doesn't work is preselecting the right index in the WebUI combobox when opening options. I'm trying to see if i can figure a way to do it in preferences.html but i don't know html and i'm not succeeding so far.

Reply:
>Anyway, what doesn't work is preselecting the right index in the WebUI combobox when opening options.

https://github.com/thalieht/qBittorrent/blob/e741e353618a90915fc1352830787402193c9757/src/webui/www/private/views/preferences.html#L1626
Here is affected place. Not sure if you can select `option` by `value` attribute value. Then you have to perform mapping manually.

Reply:
Thanks.